File: solution/0507____Perfect_Number.c
#include <stdbool.h>
#include <math.h>


bool _1(int num){
        if(num <= 1)
                return false;
        int root = sqrt(num);
        int sum_of_divisors = 1;
        for(int i = 2; i < root; ++i)
                if(num % i == 0)
                        sum_of_divisors += i + num / i;
        if(root * root == num)
                sum_of_divisors -= root;
        return sum_of_divisors == num;
}


bool checkPerfectNumber(int num){
        return _1(num);
}

int main(){
        checkPerfectNumber(28);
        return 0;
}

File: solution/0507____Perfect Number.go
package main

import "math"

// The worst of the journey is still to come

// perf: 4ms, 40%
func checkPerfectNumber(num int) bool {
	divisors := make(map[int]struct{})
	for i, temp := 1, int(math.Sqrt(float64(num))); i <= temp; i += 1 {
		if num%i == 0 {
			divisors[i] = struct{}{}
			divisors[num/i] = struct{}{}
		}
	}

	sum := 0
	for i := range divisors {
		if i != num {
			sum += i
		}
	}
	return sum == num
}

func main() {
	checkPerfectNumber(1)
	checkPerfectNumber(28)
}
